HW8-DBs:

olgakuzmins opened the pull request:
![HW8-DBs](https://github.com/user-attachments/assets/80def5da-2397-4538-b19f-afa85c86d19f)

olgakuzmins committed:
HW8-DBs:

AtticSiol commented on src/main/resources/database.properties.origin:
```diff
@@ -0,0 +1,4 @@
+driver=
```
add value

olgakuzmins replied:
added

AtticSiol commented on src/main/resources/database.properties.origin:
```diff
@@ -0,0 +1,4 @@
+driver=
+url=
```
add value

olgakuzmins replied:
added

AtticSiol commented:
add init sql script file into resources folder

AtticSiol commented on src/main/java/com/kuzmins/dao/UserDAO.java:
```diff
@@ -0,0 +1,59 @@
+import com.kuzmins.config.ConnectionConfig;
+import com.kuzmins.model.User;
+
+import java.sql.*;
```
do not use squashed imports. import each class separately, it can be configured int Intellij Idea settings

olgakuzmins replied:
fixed

olgakuzmins committed:
HW8-DBs: changes after review
- added file init.sql
- added information to properties
- described SQL imports

olgakuzmins commented:
> add init sql script file into resources folder

added

olgakuzmins committed:
HW8-DBs: changes after review
- added file init.sql
- added information to properties
- described SQL imports

AtticSiol commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,16 @@
+create table Users(
```
use `user_info` or `user_data` instead of `Users`: 1) lower case 2) simple `user` is reserved keyword in dbs

olgakuzmins replied:
renamed tables

AtticSiol commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,16 @@
+create type ticket_type as enum('DAY', 'WEEK', 'MONTH', 'YEAR');
+
+
+create table Tickets(
```
it is good practice to use name of tables in lower case

olgakuzmins replied:
fixed

AtticSiol commented on src/main/java/com/kuzmins/service/TicketService.java:
```diff
@@ -79,93 +79,43 @@ public List<Ticket> returnTicketsBySector(Sector sector) {
 
-        //delete(T element)
-        customHashSet.delete(ticket1);
+        User user3 = userDAO.fetchUserById(matt.getId());
```
add one function: 'Fetch Tickets by ID and user ID' in one command

olgakuzmins replied:
I'm not sure that I understand the comment)
Do you mean to write a method fetchTicketsByUserId(UUID userId) call just to test, how does it work?

Or do I need to add a new method into ticketDAO?

AtticSiol replied:
yes, it is about new complex method: 
```java
public Ticket fetchTicketByIdAndUserId(UUID id, UUID userId) {
...
...prepareStatement("SELECT * FROM ticket_info JOIN user_info ON user_info.id = ticket_info.user_id WHERE ticket_info.id=? and user_info.id=?")
...
```

AtticSiol commented on src/main/java/com/kuzmins/dao/TicketDAO.java:
```diff
@@ -0,0 +1,110 @@
+    public void saveTicket(Ticket ticket) {
+        try {
+            PreparedStatement preparedStatement =
+                    connection.prepareStatement("INSERT INTO tickets (id, ticket_type, creation_date) VALUES (?, ?::ticket_type, ?)");
```
perfectly will be to convert all queries (`"INSERT INTO tickets (id, ticket_type, creation_date) VALUES (?, ?::ticket_type, ?)"`) in static variables to create them once with class instead of creating with each time of method call

olgakuzmins replied:
converted

olgakuzmins replied:
added method fetchTicketByIdAndUserId(UUID id, UUID userId)

olgakuzmins committed:
HW8-DBs
- renamed tables to lower case
- converted queries into static variables

AtticSiol commented on src/main/java/com/kuzmins/dao/TicketDAO.java:
```diff
@@ -0,0 +1,117 @@
+import com.kuzmins.model.TicketType;
+import com.kuzmins.model.User;
+
+import java.sql.*;
```
avoid squashed imports

olgakuzmins replied:
added all imports

AtticSiol commented on src/main/java/com/kuzmins/dao/UserDAO.java:
```diff
@@ -0,0 +1,68 @@
+
+public class UserDAO {
+
+    private final static Connection CONNECTION = ConnectionConfig.getConnection();
```
open (and close) connection and resources for each method, do not use one opened connection for class level.
You can do it:
```java
        try {
            Connection conn= ConnectionConfig.getConnection();
            PreparedStatement preparedStatement =conn.prepareStatement(INSERT_USER_QUERY);
            ...
        } catch (SQLException e) {
           ...
        } finally {
            try {
                preparedStatement.close();
                conn.close();
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        }
```
or in `try-with-resources` construction:
```java
  try (Connection connection = ConnectionConfig.getConnection();
             PreparedStatement statement = connection.prepareStatement(sqlCommand)) {
             ...
        } catch (SQLException e) {
            System.out.println(e);
        } 
```

olgakuzmins replied:
applied try-with-resources construction

olgakuzmins committed:
HW8-DBs
- added method fetchTicketByIdAndUserId(UUID id, UUID userId)
- avoided squashed imports
- applied try-with-resources construction for Connection and Preparestatement
